File: android/app/src/main/java/com/opencontinuity/core/protocol/Protocol.kt
package com.opencontinuity.core.protocol

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement

/**
 * Protocol version for compatibility checking
 */
const val PROTOCOL_VERSION = "2.0.0"

/**
 * JSON serializer configuration
 */
val protocolJson = Json {
    ignoreUnknownKeys = true
    encodeDefaults = true
    isLenient = true
    prettyPrint = false
}

/**
 * Base message envelope for all protocol messages
 */
@Serializable
data class ProtocolMessage(
    val type: MessageType,
    val payload: JsonElement,
    val timestamp: Long = System.currentTimeMillis(),
    val messageId: String = generateMessageId(),
    val signature: String? = null
)

/**
 * All supported message types
 */
@Serializable
enum class MessageType {
    // Connection & Pairing
    @SerialName("handshake") HANDSHAKE,
    @SerialName("handshake_response") HANDSHAKE_RESPONSE,
    @SerialName("pairing_request") PAIRING_REQUEST,
    @SerialName("pairing_response") PAIRING_RESPONSE,
    @SerialName("pairing_complete") PAIRING_COMPLETE,
    @SerialName("heartbeat") HEARTBEAT,
    @SerialName("heartbeat_ack") HEARTBEAT_ACK,
    @SerialName("disconnect") DISCONNECT,

    // Clipboard
    @SerialName("clipboard_sync") CLIPBOARD_SYNC,
    @SerialName("clipboard_ack") CLIPBOARD_ACK,

    // File Transfer
    @SerialName("file_transfer_request") FILE_TRANSFER_REQUEST,
    @SerialName("file_transfer_accept") FILE_TRANSFER_ACCEPT,
    @SerialName("file_transfer_reject") FILE_TRANSFER_REJECT,
    @SerialName("file_transfer_progress") FILE_TRANSFER_PROGRESS,
    @SerialName("file_transfer_complete") FILE_TRANSFER_COMPLETE,
    @SerialName("file_transfer_error") FILE_TRANSFER_ERROR,

    // Notifications
    @SerialName("notification_post") NOTIFICATION_POST,
    @SerialName("notification_dismiss") NOTIFICATION_DISMISS,
    @SerialName("notification_action") NOTIFICATION_ACTION,

    // SMS
    @SerialName("sms_conversations") SMS_CONVERSATIONS,
    @SerialName("sms_messages") SMS_MESSAGES,
    @SerialName("sms_send") SMS_SEND,
    @SerialName("sms_send_result") SMS_SEND_RESULT,
    @SerialName("sms_received") SMS_RECEIVED,

    // Camera / Screen Mirror
    @SerialName("stream_start") STREAM_START,
    @SerialName("stream_stop") STREAM_STOP,
    @SerialName("stream_offer") STREAM_OFFER,
    @SerialName("stream_answer") STREAM_ANSWER,
    @SerialName("stream_ice_candidate") STREAM_ICE_CANDIDATE,

    // Battery & Status
    @SerialName("device_status") DEVICE_STATUS,
    @SerialName("battery_status") BATTERY_STATUS,

    // Input Control
    @SerialName("input_event") INPUT_EVENT,
    @SerialName("touchpad_event") TOUCHPAD_EVENT,

    // File Chunk Transfer
    @SerialName("file_chunk") FILE_CHUNK,

    // Cross-Device Drag and Drop
    @SerialName("drag_file_start") DRAG_FILE_START,
    @SerialName("drag_file_cancel") DRAG_FILE_CANCEL,
    @SerialName("drag_file_drop") DRAG_FILE_DROP,

    // Session restore
    @SerialName("session_restore") SESSION_RESTORE,
    @SerialName("session_restore_ack") SESSION_RESTORE_ACK,

    // Screenshot Sync
    @SerialName("screenshot_available") SCREENSHOT_AVAILABLE,
    @SerialName("screenshot_request") SCREENSHOT_REQUEST,

    // Error
    @SerialName("error") ERROR
}

// ============== Pairing Messages ==============

@Serializable
data class HandshakePayload(
    val deviceName: String,
    val deviceType: DeviceType,
    val protocolVersion: String = PROTOCOL_VERSION,
    val publicKey: String,
    val features: List<String> = emptyList(),
    /** Granular capability flags for dynamic feature negotiation */
    val capabilities: DeviceCapabilities? = null,
    val deviceId: String? = null
)

@Serializable
data class DeviceCapabilities(
    val clipboard: Boolean = true,
    val input: Boolean = true,
    val fileTransfer: Boolean = true,
    val dragDrop: Boolean = true,
    val notifications: Boolean = true,
    val sms: Boolean = true,
    val camera: Boolean = true,
    val screenMirror: Boolean = true,
    val battery: Boolean = true
)

@Serializable
data class HandshakeResponsePayload(
    val accepted: Boolean,
    val deviceName: String,
    val deviceType: DeviceType,
    val protocolVersion: String = PROTOCOL_VERSION,
    val publicKey: String,
    val sessionToken: String? = null,
    val features: List<String> = emptyList()
)

@Serializable
data class PairingRequestPayload(
    val pairingCode: String,
    val deviceName: String,
    val publicKey: String
)

@Serializable
data class PairingResponsePayload(
    val success: Boolean,
    val sessionToken: String? = null,
    val errorMessage: String? = null
)

@Serializable
enum class DeviceType {
    @SerialName("android") ANDROID,
    @SerialName("windows") WINDOWS,
    @SerialName("unknown") UNKNOWN
}

// ============== Clipboard Messages ==============

@Serializable
data class ClipboardSyncPayload(
    val contentType: ClipboardContentType,
    val textContent: String? = null,
    val htmlContent: String? = null,
    val imageBase64: String? = null,
    val imageMimeType: String? = null,
    /** Source device ID — receivers skip own messages to prevent echo */
    val deviceId: String? = null,
    /** SHA-256 hex of content for cross-device deduplication */
    val contentHash: String? = null
)

@Serializable
enum class ClipboardContentType {
    @SerialName("text") TEXT,
    @SerialName("html") HTML,
    @SerialName("image") IMAGE
}

// ============== File Transfer Messages ==============

@Serializable
data class FileTransferRequestPayload(
    val transferId: String,
    val fileName: String,
    val fileSize: Long,
    val mimeType: String,
    val checksum: String? = null
)

@Serializable
data class FileTransferProgressPayload(
    val transferId: String,
    val bytesTransferred: Long,
    val totalBytes: Long,
    val progress: Float
)

@Serializable
data class FileTransferCompletePayload(
    val transferId: String,
    val success: Boolean,
    val filePath: String? = null,
    val errorMessage: String? = null
)

// ============== File Chunk Messages ==============

@Serializable
data class FileChunkPayload(
    val transferId: String,
    /** 0-based chunk index */
    val sequence: Int,
    val totalChunks: Int,
    /** Base64-encoded raw bytes */
    val data: String,
    /** SHA-256 hex of raw chunk bytes for per-chunk integrity */
    val checksum: String
)

// ============== Cross-Device Drag and Drop ==============

@Serializable
data class DragFileStartPayload(
    val dragId: String,
    val sourceDeviceId: String,
    val fileName: String,
    val fileSize: Long,
    val mimeType: String,
    /** Normalized 0–1 screen-edge position where drag exited */
    val edgeX: Float,
    val edgeY: Float
)

@Serializable
data class DragFileDropPayload(
    val dragId: String,
    val targetDeviceId: String,
    val accepted: Boolean,
    val errorMessage: String? = null
)

// ============== Session Restore ==============

@Serializable
data class SessionRestorePayload(
    val sessionToken: String,
    val deviceId: String,
    val deviceName: String
)

// ============== Notification Messages ==============

@Serializable
data class NotificationPayload(
    val notificationId: String,
    val packageName: String,
    val appName: String,
    val title: String?,
    val text: String?,
    val subText: String? = null,
    val iconBase64: String? = null,
    val timestamp: Long,
    val actions: List<NotificationAction> = emptyList()
)

@Serializable
data class NotificationAction(
    val actionId: String,
    val title: String
)

// ============== SMS Messages ==============

@Serializable
data class SmsConversation(
    val threadId: Long,
    val address: String,
    val contactName: String?,
    val snippet: String?,
    val timestamp: Long,
    val unreadCount: Int
)

@Serializable
data class SmsMessage(
    val id: Long,
    val threadId: Long,
    val address: String,
    val body: String,
    val timestamp: Long,
    val isIncoming: Boolean,
    val isRead: Boolean
)

@Serializable
data class SmsSendPayload(
    val address: String,
    val body: String,
    val requestId: String
)

@Serializable
data class SmsSendResultPayload(
    val requestId: String,
    val success: Boolean,
    val errorMessage: String? = null
)

// ============== Device Status Messages ==============

@Serializable
data class BatteryStatusPayload(
    val level: Int,
    val isCharging: Boolean,
    val chargeType: String,
    val temperature: Float,
    val health: String
)

@Serializable
data class DeviceStatusPayload(
    val deviceName: String,
    val battery: BatteryStatusPayload,
    val storageUsedBytes: Long,
    val storageTotalBytes: Long,
    val wifiConnected: Boolean,
    val wifiSsid: String?
)

// ============== Input Control Messages ==============

@Serializable
data class InputEventPayload(
    val eventType: InputEventType,
    val x: Float? = null,
    val y: Float? = null,
    val keyCode: Int? = null,
    val text: String? = null,
    val scrollDelta: Float? = null
)

@Serializable
enum class InputEventType {
    @SerialName("tap") TAP,
    @SerialName("long_press") LONG_PRESS,
    @SerialName("swipe") SWIPE,
    @SerialName("scroll") SCROLL,
    @SerialName("key") KEY,
    @SerialName("text") TEXT
}

@Serializable
data class TouchpadEventPayload(
    val eventType: TouchpadEventType,
    val deltaX: Float? = null,
    val deltaY: Float? = null,
    val fingers: Int = 1,
    val scrollDelta: Float? = null
)

@Serializable
enum class TouchpadEventType {
    @SerialName("move") MOVE,
    @SerialName("click") CLICK,
    @SerialName("right_click") RIGHT_CLICK,
    @SerialName("scroll") SCROLL,
    @SerialName("drag_start") DRAG_START,
    @SerialName("drag_end") DRAG_END
}

// ============== Error Messages ==============

@Serializable
data class ErrorPayload(
    val code: String,
    val message: String,
    val details: String? = null
)

// ============== Utility Functions ==============

private fun generateMessageId(): String {
    return java.util.UUID.randomUUID().toString()
}

File: windows/src/shared/protocol.ts
/**
 * Shared Protocol Types for OpenContinuity
 * These types define the message structure between Android and Windows apps
 * Keep in sync with android/app/.../core/protocol/Protocol.kt
 * Keep in sync with /shared/protocol.ts (root)
 */

export enum MessageType {
  // Connection & Pairing
  HANDSHAKE = 'handshake',
  HANDSHAKE_RESPONSE = 'handshake_response',
  PAIRING_REQUEST = 'pairing_request',
  PAIRING_RESPONSE = 'pairing_response',
  PAIRING_COMPLETE = 'pairing_complete',
  HEARTBEAT = 'heartbeat',
  HEARTBEAT_ACK = 'heartbeat_ack',
  DISCONNECT = 'disconnect',

  // Clipboard
  CLIPBOARD_SYNC = 'clipboard_sync',
  CLIPBOARD_ACK = 'clipboard_ack',

  // File Transfer
  FILE_TRANSFER_REQUEST = 'file_transfer_request',
  FILE_TRANSFER_ACCEPT = 'file_transfer_accept',
  FILE_TRANSFER_REJECT = 'file_transfer_reject',
  FILE_TRANSFER_PROGRESS = 'file_transfer_progress',
  FILE_TRANSFER_COMPLETE = 'file_transfer_complete',
  FILE_TRANSFER_ERROR = 'file_transfer_error',

  // Notifications
  NOTIFICATION_POST = 'notification_post',
  NOTIFICATION_DISMISS = 'notification_dismiss',
  NOTIFICATION_ACTION = 'notification_action',

  // SMS
  SMS_CONVERSATIONS = 'sms_conversations',
  SMS_MESSAGES = 'sms_messages',
  SMS_SEND = 'sms_send',
  SMS_SEND_RESULT = 'sms_send_result',
  SMS_RECEIVED = 'sms_received',

  // Camera / Screen Mirror
  STREAM_START = 'stream_start',
  STREAM_STOP = 'stream_stop',
  STREAM_OFFER = 'stream_offer',
  STREAM_ANSWER = 'stream_answer',
  STREAM_ICE_CANDIDATE = 'stream_ice_candidate',

  // Battery & Status
  DEVICE_STATUS = 'device_status',
  BATTERY_STATUS = 'battery_status',

  // Input Control
  INPUT_EVENT = 'input_event',
  TOUCHPAD_EVENT = 'touchpad_event',

  // File Chunk Transfer (AirDrop-style chunked data)
  FILE_CHUNK = 'file_chunk',

  // Cross-Device Drag and Drop
  DRAG_FILE_START = 'drag_file_start',
  DRAG_FILE_CANCEL = 'drag_file_cancel',
  DRAG_FILE_DROP = 'drag_file_drop',

  // Session
  SESSION_RESTORE = 'session_restore',
  SESSION_RESTORE_ACK = 'session_restore_ack',

  // Screenshot Sync
  SCREENSHOT_AVAILABLE = 'screenshot_available',
  SCREENSHOT_REQUEST = 'screenshot_request',

  // Note Maker Sync
  NOTE_SYNC = 'note_sync',

  // Error
  ERROR = 'error'
}

export interface ProtocolMessage<T = any> {
  type: MessageType;
  payload: T;
  timestamp: number;
  messageId: string;
  signature?: string;
}

// Alias for compatibility
export type Message<T = any> = ProtocolMessage<T>;

// ============== Pairing Messages ==============

export type DeviceType = 'android' | 'windows' | 'unknown';

export interface DeviceCapabilities {
  clipboard: boolean;
  input: boolean;
  fileTransfer: boolean;
  dragDrop: boolean;
  notifications: boolean;
  sms: boolean;
  camera: boolean;
  screenMirror: boolean;
  battery: boolean;
}

export interface HandshakePayload {
  deviceName: string;
  deviceType: DeviceType;
  protocolVersion: string;
  publicKey: string;
  features: string[];
  capabilities?: DeviceCapabilities;
  deviceId?: string;
}

export interface HandshakeResponsePayload {
  accepted: boolean;
  deviceName: string;
  deviceType: DeviceType;
  protocolVersion: string;
  publicKey: string;
  sessionToken?: string;
  features: string[];
}

export interface PairingRequestPayload {
  pairingCode: string;
  deviceName: string;
  publicKey: string;
}

export interface PairingResponsePayload {
  success: boolean;
  sessionToken?: string;
  errorMessage?: string;
}

// ============== Clipboard Messages ==============

export type ClipboardContentType = 'text' | 'html' | 'image';

export interface ClipboardSyncPayload {
  contentType: ClipboardContentType;
  textContent?: string;
  htmlContent?: string;
  imageBase64?: string;
  imageMimeType?: string;
  deviceId?: string;
  contentHash?: string;
}

// ============== File Transfer Messages ==============

export interface FileTransferRequestPayload {
  transferId: string;
  fileName: string;
  fileSize: number;
  mimeType: string;
  checksum?: string;
}

export interface FileTransferProgressPayload {
  transferId: string;
  bytesTransferred: number;
  totalBytes: number;
  progress: number;
}

export interface FileTransferCompletePayload {
  transferId: string;
  success: boolean;
  filePath?: string;
  errorMessage?: string;
}

export interface FileChunkPayload {
  transferId: string;
  sequence: number;
  totalChunks: number;
  data: string; // base64
  checksum: string;
}

export interface DragFileStartPayload {
  dragId: string;
  sourceDeviceId: string;
  fileName: string;
  fileSize: number;
  mimeType: string;
  edgeX: number;
  edgeY: number;
}

export interface DragFileDropPayload {
  dragId: string;
  targetDeviceId: string;
  accepted: boolean;
  errorMessage?: string;
}

export interface SessionRestorePayload {
  sessionToken: string;
  deviceId: string;
  deviceName: string;
}

// ============== Notification Messages ==============

export interface NotificationAction {
  actionId: string;
  title: string;
}

export interface NotificationPayload {
  notificationId: string;
  packageName: string;
  appName: string;
  title?: string;
  text?: string;
  subText?: string;
  iconBase64?: string;
  timestamp: number;
  actions: NotificationAction[];
}

// ============== SMS Messages ==============

export interface SmsConversation {
  threadId: number;
  address: string;
  contactName?: string;
  snippet?: string;
  timestamp: number;
  unreadCount: number;
}

export interface SmsMessage {
  id: number;
  threadId: number;
  address: string;
  body: string;
  timestamp: number;
  isIncoming: boolean;
  isRead: boolean;
}

export interface SmsSendPayload {
  address: string;
  body: string;
  requestId: string;
}

export interface SmsSendResultPayload {
  requestId: string;
  success: boolean;
  errorMessage?: string;
}

// ============== Device Status Messages ==============

export interface BatteryStatusPayload {
  level: number;
  isCharging: boolean;
  chargeType: string;
  temperature: number;
  health: string;
}

export interface DeviceStatusPayload {
  deviceName: string;
  battery: BatteryStatusPayload;
  storageUsedBytes: number;
  storageTotalBytes: number;
  wifiConnected: boolean;
  wifiSsid?: string;
}

// ============== Input Control Messages ==============

export type InputEventType = 'tap' | 'long_press' | 'swipe' | 'scroll' | 'key' | 'text';

export interface InputEventPayload {
  eventType: InputEventType;
  x?: number;
  y?: number;
  keyCode?: number;
  text?: string;
  scrollDelta?: number;
}

export type TouchpadEventType = 'move' | 'click' | 'right_click' | 'scroll' | 'drag_start' | 'drag_end';

export interface TouchpadEventPayload {
  eventType: TouchpadEventType;
  deltaX?: number;
  deltaY?: number;
  fingers: number;
  scrollDelta?: number;
}

// ============== Stream Messages ==============

export interface StreamStartPayload {
  streamType: 'screen' | 'camera';
  width?: number;
  height?: number;
  fps?: number;
  bitrate?: number;
}

export interface StreamStopPayload {
  streamType: 'screen' | 'camera';
}

// ============== Note Maker Sync ==============

export type NoteTool = 'pen' | 'eraser' | 'cursor';
export type NoteSyncAction = 'stroke' | 'clear' | 'pan' | 'zoom';

export interface NotePoint {
  x: number;
  y: number;
}

export interface NoteSyncPayload {
  action: NoteSyncAction;
  tool: NoteTool;
  color: string;
  thickness: number;
  points: NotePoint[];
  panX?: number;
  panY?: number;
  zoom?: number;
}

// ============== Error Messages ==============

export interface ErrorPayload {
  code: string;
  message: string;
  details?: string;
}

// ============== Utility ==============

export const PROTOCOL_VERSION = '2.0.0';

export function generateMessageId(): string {
  return `${Date.now()}-${Math.random().toString(36).substr(2, 9)}`;
}

export function createMessage<T>(type: MessageType, payload: T): ProtocolMessage<T> {
  return {
    type,
    payload,
    timestamp: Date.now(),
    messageId: generateMessageId()
  };
}

export const protocolJson = {
  parse: (text: string) => JSON.parse(text) as ProtocolMessage,
  stringify: (message: ProtocolMessage) => JSON.stringify(message)
};
